### DISCUSSION TITLE
Wormholes

### DISCUSSION THREAD

**Discussion by IsabellaB622** [godotengine/godot-proposals#4982] [3D]
I believe it would be beneficial to the Godot community to add a "wormhole" node
wormhole nodes would allow physics objects (Kinematic bodies and rigid bodies for example) to be teleported to a new area of the same scene (or if possible to a new scene) wormhole's shape could be any of the 3D or 2D primitives. i think this would benefit not only people making "portal clones" but anyone who wants to utilize Non-euclidean geometry such as those who want a building to have a small footprint but still be usable on the inside. wormholes would preserve the body's direction and velocity (an object that enters a wormhole pointed south (body going north at 1m/s) that is linked to a wormhole pointed north would exit the wormhole going south at 1 m/s) wormholes could be customly textured with a "wormhole texture" object (which would link to another wormhole node) and would display that which would be visible from the other side. while this can be implemented now in Godot implementations are often buggy, unreliable, and difficult to implement for the average user (this thought is based off of YouTube videos of people trying to implement this feature that I've seen). 

Edit:  (an object that enters a wormhole pointed south (body going north at 1m/s) that is linked to a wormhole pointed north would exit the wormhole going NORTH at 1 m/s not south) likewise   (an object that enters a wormhole pointed south (body going north at 1m/s) that is linked to a wormhole pointed east would exit the wormhole going east at 1 m/s)

**Comment by Zireael07** [comment]
How does this differ from just adding a new Node2D/3D or RemoteTransform2D/3D?

**Reply by IsabellaB622** [reply]
Maybe I'm just misunderstanding the question but can you elaborate some? I fail to see how a Node2D/3D or RemoteTransform2D/3D node provides ANY of the functionality described.

**Reply by Zireael07** [reply]
It doesn't provide the functionality described but the way I'd implement a wormmhole would be just an Area2D/3D that sends stuff to some Node/RemoteTransform...

**Comment by seppoday** [comment]
Godot Engine is kinda not build like that. You have basic nodes that can be easly tweeked with a little bit of programming. Going that route we would need nodes for every possible game feature. You can easly achieve wormhole effect with 1-2 nodes and few lines of code.

**Reply by IsabellaB622** [reply]
"You can easly achieve wormhole effect with 1-2 nodes and few lines of code."
explain? (please release any sample code under the MIT license thanks)

**Comment by eon-s** [comment]
Even if it is super easy to implement, a cheap built-in teleporting capability for rigids in the server could be nice to have.

Something like (the defunct) Nape Physics' "portal", which also affect the scale of the shapes
https://joecreates.github.io/napephys/samples.html
(you'll need a flash player to see it)

**Reply by Zireael07** [reply]
Ah, for rigidbodies ... a use case I failed to consider...

**Comment by elvisish** [comment]
Here's how to transform the player to the relative position, velocity and rotation of another teleporter. I originally put it together in Unity based on a video and a comment I found buried on Unity Answers and converted a few months ago to GDScript:

```swift
func teleport(body):
    var transform_it = transform.xform_inv(body.transform.origin)
    var transform_point = other_tele.to_global(transform_it)
    body.transform.origin = transform_point

    var rotation = Quat(global_transform.basis).inverse()
    var player_rotation = Quat(body.global_transform.basis)
    var other_rotation = Quat(other_tele.global_transform.basis)
    body.global_transform.basis = Basis(rotation * player_rotation * other_rotation)

    body.velocity = other_tele.transform.basis.xform(transform.basis.xform_inv(body.velocity))
```
